src/MgvOfferTaking.sol

```diff
@@ -191,13 +197,23 @@ abstract contract MgvOfferTaking is MgvHasOffers {
   /* `internalMarketOrder` works recursively. Going downward, each successive offer is executed until the market order stops (due to: volume exhausted, bad price, or empty book). Then the [reentrancy lock is lifted](#internalMarketOrder/liftReentrancy). Going upward, each offer's `maker` contract is called again with its remaining gas and given the chance to update its offers on the book. */
   function internalMarketOrder(Pair storage pair, MultiOrder memory mor, MgvLib.SingleOrder memory sor) internal {
     unchecked {
-      if (mor.fillVolume == 0 || sor.offerId == 0 || !sor.offer.tick().better(mor.maxTick)) {
+      if (
+        mor.fillVolume == 0 || sor.offerId == 0 || mor.maxRecursionDepth == 0 || !sor.offer.tick().better(mor.maxTick)
+      ) {
         return endInternalMarketOrder(pair, mor, sor);
       }
 
       /* Load additional information about the offer. */
       sor.offerDetail = pair.offerData[sor.offerId].detail;
 
+      mor.gasreqLeft += sor.offerDetail.gasreq();
+
+      if (gasleft() < mor.gasreqLeft + sor.local.offer_gasbase()) {
+        return endInternalMarketOrder(pair, mor, sor);
+      }
+
+      mor.maxRecursionDepth--;
+
```

Comment:
Ok I understand why the function `endInternalMarketOrder` was created!

Since moving to ticks, we have 72 bits available in `offer`. I think moving `gasreq` from `offerDetail` to `offer` make sense. It would reduce gas use in some situations, and mainly here it may remove one call to `endInternalMarketOrder` because we could just add `|| gasleft() < ...` to the big `if` at the beginning of `internalMarketOrder`. With the proviso that updating `mor.gasreqLeft` early is not a problem.

By removing one call to it we would then be able to inline it.

Reply:
Are you OK with adding this as a task to our general PR, so that this PR can be completed?

Reply:
ok!